enterprise/tailnet/coordinator.go

```diff
@@ -0,0 +1,507 @@
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Update the node of this client in our in-memory map. If an agent entirely
+	// shuts down and reconnects, it needs to be aware of all clients attempting
+	// to establish connections.
+	c.nodes[id] = &node
+
+	// Write the new node from this client to the actively connected agent.
+	err = c.writeNodeToAgent(agent, &node)
+	if err != nil {
+		return xerrors.Errorf("write node to agent: %w", err)
+	}
+
+	return nil
+}
+
+func (c *haCoordinator) writeNodeToAgent(agent uuid.UUID, node *agpl.Node) error {
+	agentSocket, ok := c.agentSockets[agent]
+	if !ok {
+		// If we don't own the agent locally, send it over pubsub to a node that
+		// owns the agent.
+		err := c.publishNodesToAgent(agent, []*agpl.Node{node})
+		if err != nil {
+			return xerrors.Errorf("publish node to agent")
+		}
+		return nil
+	}
+
+	// Write the new node from this client to the actively
+	// connected agent.
+	data, err := json.Marshal([]*agpl.Node{node})
+	if err != nil {
+		return xerrors.Errorf("marshal nodes: %w", err)
+	}
+
+	_, err = agentSocket.Write(data)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		return xerrors.Errorf("write json: %w", err)
+	}
+	return nil
+}
+
+// ServeAgent accepts a WebSocket connection to an agent that listens to
+// incoming connections and publishes node updates.
+func (c *haCoordinator) ServeAgent(conn net.Conn, id uuid.UUID) error {
+	c.mutex.Lock()
+
+	// Tell clients on other instances to send a callmemaybe to us.
+	err := c.publishAgentHello(id)
+	if err != nil {
+		return xerrors.Errorf("publish agent hello: %w", err)
+	}
+
+	// Publish all nodes on this instance that want to connect to this agent.
+	nodes := c.nodesSubscribedToAgent(id)
+	if len(nodes) > 0 {
+		data, err := json.Marshal(nodes)
+		if err != nil {
+			c.mutex.Unlock()
+			return xerrors.Errorf("marshal json: %w", err)
+		}
+		_, err = conn.Write(data)
+		if err != nil {
+			c.mutex.Unlock()
+			return xerrors.Errorf("write nodes: %w", err)
+		}
+	}
+
+	// If an old agent socket is connected, we close it
+	// to avoid any leaks. This shouldn't ever occur because
+	// we expect one agent to be running.
+	oldAgentSocket, ok := c.agentSockets[id]
+	if ok {
+		_ = oldAgentSocket.Close()
+	}
+	c.agentSockets[id] = conn
+	c.mutex.Unlock()
+	defer func() {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		delete(c.agentSockets, id)
+		delete(c.nodes, id)
+	}()
+
+	decoder := json.NewDecoder(conn)
+	for {
+		node, err := c.hangleAgentUpdate(id, decoder)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return xerrors.Errorf("handle next agent message: %w", err)
+		}
+
+		err = c.publishAgentToNodes(id, node)
+		if err != nil {
+			return xerrors.Errorf("publish agent to nodes: %w", err)
+		}
+	}
+}
+
+func (c *haCoordinator) nodesSubscribedToAgent(agentID uuid.UUID) []*agpl.Node {
+	sockets, ok := c.agentToConnectionSockets[agentID]
```

Comment:
this pattern that has an implicit dependence on callers holding a global mutex scares me a bit.